Add petite-vue plugins support

ws-rush opened the pull request:
I am so excited, this is my first pull request on github, I dont know typescript, and I wish this pull request accepted

I did this commit to distribute custom directives as packages then use them with `petite-vue`, I also added example use in README file, thats it:

### Use Plugins

You can write custome directive then distrbute it as a pacage, then add it to create vue, like:
```html
<div v-scope="{counter: 0}" v-log="inside petite-vue scope">
  <button @click="counter++">increase</button>
</div>

<script type="module">
  import log from './log'
  import { createApp } from 'peteite-vue'
  createApp().use(log).mount()
</script>
```

A plugin code similar to vue plugins code:

```js
// inside log.js plugin file
export default {
  install: (app, options) => {
    app.directive('log', ({exp}) => {
      console.log(exp)
    })
  }
}
```

ws-rush committed:
Add petite-vue plugins support

Commit:
Add `$root` refer to component root element

Some times like in dialogs we need access to root component from child nodes, also it is a way to access root element from `v-scope` . the next snippet from another PR to expose `$el` to scope, but this PR can solve the same problem also

```html
<textarea
  v-scope="{width: $root.offsetWidth, height: $root.offsetHeight}"
  @click="width = $el.offsetWidth; height = $el.offsetHeight;"
>
{{ width }} &times; {{ height }}
</textarea>
```

Commit:
Change scope of ref with `v-scope`

ws-rushx committed:
Change project name

ws-rushx committed:
typo

ws-rushx committed:
Change package name and version

ws-rushx committed:
Expose $el to v-scope

ws-rushx committed:
Update package and Readme

ws-rushx committed:
Fix refrences

ws-rushx committed:
Change release

ws-rushx committed:
Support <template> in

ws-rushx committed:
Rename project

sqllyw commented:
what is the advantage of use() over directive() ? example:

````
createApp().directive('log',log).mount() vs 
createApp().use(log).mount()
````

lucafabbian committed:
Added `watchEffect`

ws-rush committed:
Merge pull request #1 from lucafabbian/main

ws-rushx committed:
fix: fix pico-vue dependcies to 3.2.27

ws-rushx committed:
deps: update typescript dependcy to 5.4.5

ws-rushx committed:
dependcy: upgrade vite to 5.2.13

ws-rushx committed:
depndcy: uupdate @vue/reactivity to 3.2.45

ws-rushx committed:
dependcy: fix @vue@reactivity version

ws-rushx committed:
release: pico-vue@1.0.5

ws-rush committed:
change: rename project to pocket-vue

ws-rush committed:
fix: update commits example to use working API endpoint
   - Replace deprecated vuejs/vue-next with microsoft/vscode repository
   - Fix branch names from wildcards to specific release branches

ws-rush committed:
Merge pull request #5 from ws-rush/fix-exmaple

fix: update commits example to use working API endpoint

ws-rush committed:
test: add comprehensive test suite and testing guide

  - Add complete Vitest configuration with Happy DOM environment
  - Create 119 tests covering all core modules and directives
  - Achieve 82.84% test coverage across the entire codebase
  - Add tests for app creation, mounting, and lifecycle
  - Test all built-in directives (v-bind, v-on, v-model, v-show, v-text, v-html, v-effect)
  - Include comprehensive reactivity and async testing patterns
  - Add error handling and edge case coverage
  - Create integration tests for component-like behavior
  - Add tests for utilities, scheduler, and DOM walking functionality
  - Write comprehensive testing guide with best practices
  - Document common testing patterns and examples

ws-rush committed:
fix: types for { moduleResolution: Node16 }

ws-rush committed:
chore: update dependencies and build configuration to modern standards

  - Add ES module support with type: module
  - Upgrade Vue dependencies from 3.2.45 to 3.5.21
  - Update dev dependencies to latest stable versions

ws-rush committed:
Merge pull request #6 from ws-rush/upgrade-deps

chore: update dependencies and build configuration to modern standards

ws-rush committed:
fix: build command

ws-rush committed:
Merge pull request #7 from ws-rush/fix--build-command

fix: build command

ws-rush committed:
change: convert 'range' automaticlly to number without need .number modifire as in number input

ws-rush committed:
chore: typo

ws-rush committed:
fix: todomvc example

ws-rush committed:
docs: update CDN links and version information

  - Update production CDN URLs from version 1.0.3 to 1.0.5
  - Add comprehensive CDN links section with jsDelivr and unpkg options
  - Provide direct links for both global and ESM builds across CDN providers

ws-rush committed:
fix: v-if nested in v-for template tag

  - Use parentNode fallback when parentElement is null
  - Add safety check to prevent runtime errors
  - Handle template tag scenarios in v-for loops
  - Prevents crashes when v-if is used inside template elements

ws-rush committed:
Merge pull request #9 from ws-rush/enhances

Enhances

ws-rush committed:
release: v1.0.0

ws-rush committed:
Merge pull request #10 from ws-rush/enhances

release: v1.0.0

ws-rush committed:
fix(v-bind): Enhance reactivity and testing reliability

This commit addresses several issues and introduces improvements across the library, primarily focusing on the `v-bind` directive's reactivity and enhancing the test environment's reliability.

**Bug Fixes:**

	- **`ReferenceError: style is not defined` in `src/directives/bind.ts`:** Resolved an issue where the `style` object was not consistently defined, leading to errors in style binding.

**Refactoring & Enhancements:**

	- **`v-bind` Directive (`src/directives/bind.ts`):**
	- Removed `happy-dom`-specific conditional workarounds, restoring the code to its original, idiomatic DOM manipulation logic.
	- **Test Environment (`src/test/setup.ts`):**
	- Implemented a `happy-dom` patch for `Element.prototype.setAttribute`. This patch forces attribute updates (for `id`, `class`, `style`, `title`, `lang`, `dir`) by performing `removeAttribute` before `setAttribute` in the test environment. This ensures reliable test execution without polluting application code.
	- **Directive Integration (`src/directives/index.ts`):**
		- Formalized the inclusion of `ref`, `v-for` (`_for`), and `v-if` (`_if`) directives into `builtInDirectives`.
	- **`v-for` and `v-if` Directives (`src/directives/for.ts`, `src/directives/if.ts`):**
		- Corrected `evaluate` calls to consistently pass the `el` argument, ensuring proper evaluation context.
	- **`v-model` Directive (`src/directives/model.ts`):**
		- Improved behavior for `<select multiple>` elements by ensuring new array instances are created on update.
		- Adjusted checkbox array handling for consistency.
	- **`v-ref` Directive (`src/directives/ref.ts`):**
		- Added fallback logic to use the expression string as the ref name if the evaluated ref value is `undefined`.
	- **`evaluate` and `execute` Functions (`src/eval.ts`):**
		- Refactored `evaluate` for direct `new Function` usage and optimized `execute` with `evalCache`.
	- **`nextTick` Function (`src/scheduler.ts`):**
		- Enhanced `nextTick` to support both callback and Promise-based waiting for job completion.
	- **`createApp` Function (`src/app.ts`):**
		- Added `rootBlocks` and `scope` getters to the app instance.
		- Improved error reporting for invalid selectors.
	- **`walk` Function (`src/walk.ts`):**
		- Updated to reflect changes in `ref` directive handling and `evaluate` context.

 **Chore:**

	- **Dependency Updates:** Updated various development dependencies (`@types/node`, `happy-dom`, `typescript`, `vitest`, `terser`).
	- **Script Changes:** Renamed `test:run` to `test:once` and added a `typecheck` script in `package.json`.
	- **Configuration Update:** Updated `moduleResolution` to `bundler` in `tsconfig.json`.

	This commit significantly improves the stability and maintainability of the library's testing setup while also delivering minor bug fixes and feature enhancements.

ws-rush committed:
fix(v-bind): correctly remove attributes when value is null/undefined

The `v-bind` directive was not correctly removing attributes like `id`, `title`, `lang`, and `dir` when their bound values became `null` or `undefined`. Previously, `el.setAttribute(key, value)` would set the attribute's value to the string "null" or "undefined" instead of removing the attribute entirely.

This commit modifies the `setProp` function in `src/directives/bind.ts` to explicitly call `el.removeAttribute(key)` when the value for these specific attributes is `null` or `undefined`.

Additionally, a logical error in `__tests__/directives.test.ts` was corrected. An `expect(div?.hasAttribute("id")).toBe(false);` assertion was incorrectly placed before the `data.id` was set to `undefined`, causing a premature test failure. This assertion has been removed from its incorrect position.

ws-rush committed:
chore: clean unneeded tests

ws-rush committed:
fix: package published w/o dist

ws-rush committed:
release: v1.0.1

ws-rush committed:
fix: elemenate node spesfic code from bundles

ws-rush committed:
fix(doc): correct cdn links

ws-rush committed:
release: v1.0.2

ws-rush committed:
config: add types export

ws-rush committed:
Merge branch 'main' of https://github.com/ws-rush/pocket-vue into test-coverage

ws-rush committed:
fix: types generation

ws-rush committed:
fix: types generation

ws-rush committed:
release: v1.1.0

ws-rush committed:
release: v1.1.1

ws-rush committed:
release: v1.1.2

ws-rush committed:
Merge pull request #11 from ws-rush/test-coverage

Test coverage

ws-rush committed:
chore(deps): upgrade to vitest 4

ws-rush committed:
feat: migrate tests from happy-dom to browser mode with Playwright

- Update vitest.config.ts to use browser mode with Playwright provider
- Configure Chromium browser instance for testing
- Remove happy-dom dependency from package.json
- Simplify test setup to avoid browser API conflicts
- All tests now run in real browser environment (188/192 passing)

ws-rush committed:
fix: make tests compatible with browser mode

- Fix DocumentFragment nodeName expectation in block.test.ts
- Replace vi.doMock with direct testing approach in index.test.ts
- Fix async/await syntax in test functions
- Add tests for auto-mount behavior with and without init attribute
- All 195 tests now pass in browser mode

ws-rush committed:
feat: enhance test coverage and fix type errors

  • Add comprehensive tests for evaluate, execute, walk, for, if, model, and on directives to improve
    branch coverage from 84% to 88%
  • Fix type errors in index.test.ts, eval.test.ts, and directives.test.ts related to HTMLScriptElement,
    function parameters, and reactive object types
  • Fix bug in resolveTemplate to prevent errors when template selector doesn't match any element
  • Update test assertions and type annotations for better type safety

  This commit significantly improves the robustness and test coverage of the pocket-vue library.

ws-rush committed:
chore(tests): enhance branch test coverage

ws-rush committed:
chore(test): enhance tests quality

ws-rush committed:
chore(tests): enhance tests consistincy

ws-rush committed:
Merge pull request #14 from ws-rush/test-coverage

Test coverage

ws-rush committed:
Add DeepWiki badge to README

Added a badge for DeepWiki to the README.

google-labs-jules[bot] committed:
feat: add fileName to vite config

add fileName to vite config to fix esm module cannot be found in cdn

close #22

ws-rush committed:
Merge pull request #15 from ws-rush/fix/cdn-build-issue

Fix CDN build issue for ES modules

ws-rush committed:
Add esm.sh ESM build link to README

ws-rush committed:
refactor: whole code base

ws-rush committed:
release: v1.2.0

ws-rush committed:
release: package

ws-rush committed:
feat: Set up VitePress documentation, including new content pages and build output.

ws-rush committed:
docs: remove outdated statement about fork focus from README.md

ws-rush committed:
docs: Phase 2: Core API documentation complete

ws-rush committed:
docs: Phase 3: Directive documentation complete

ws-rush committed:
docs: Phase 4: Server-side framework integration guides complete

ws-rush committed:
docs: Phase 5: Examples and cookbook complete

ws-rush committed:
docs: Phase 6: Finalize documentation and fix build issues

ws-rush committed:
test: Enhance ref handling in `walk` and `app` tests, and improve teardown verification in `block` tests.

ws-rush committed:
docs: Update documentation, rebuild VitePress site, and modify `app.ts`.

ws-rush committed:
feat: Introduce JSR publishing support, define the `App` interface, and refine internal types and directive imports.

ws-rush committed:
release: v0.0.1

ws-rush committed:
build: Update Pico-Vue library, type definitions, and rebuild documentation.

ws-rush committed:
release: v0.0.2

ws-rush committed:
feat: automate JSR publishing with a new GitHub Actions workflow and enhance documentation with JSDoc comments.

ws-rush committed:
Merge pull request #16 from ws-rush/docs

Docs

ws-rush committed:
release: v0.0.3

ws-rush committed:
ci: fix missing dependencies in jsr-publish workflow by installing pnpm

ws-rush committed:
release: v0.0.4

ws-rush committed:
docs: fix inaccurate API docs and add $el support to all directive expressions

Code changes:
- src/eval.ts: pass $el to evaluate() generated functions so $el is
  available in all directive expressions (v-on, v-bind, v-model, etc.)
  not just v-effect. Added eval caching for performance.

Docs changes:
- Remove ref() and computed() from reactivity docs (not exported)
- Fix $el and $data docs to reflect availability in all directives
- Fix v-model true-value/false-value to use :binding syntax
- Remove non-existent provide/inject from scope-context docs
- Fix version pin from 1.2.0 to 0.0.4 in installation docs
- Fix broken this.$effect/onMounted in state-management guide
- Clarify mounted() is not an automatic hook in components docs
- Document built-in expression validation in security docs
- Add .exact modifier, system modifiers, v-on="obj" limitation
- Document key-based diffing in v-for docs
- Update api-reference with accurate exports and special properties
- Fix fetch-api, infinite-scroll, laravel examples to use @vue:mounted

262 tests pass, build clean.

ws-rush committed:
chore: switch from JSR to npm publishing with @rush/pico-vue

- Rename package from pico-vue to @rush/pico-vue with publishConfig
- Replace .github/workflows/jsr-publish.yml with npm-publish.yml
  (uses NODE_AUTH_TOKEN secret for npm registry auth)
- Remove jsr.json (no longer needed)
- Update scripts/release.cjs: remove jsr.json version bump, fix
  yarn -> pnpm commands
- Update all docs CDN links: unpkg, jsdelivr, esm.sh now reference
  @rush/pico-vue
- Update all npm install and import references in docs

Note: requires NPM_TOKEN secret to be set in GitHub repository
settings before next release tag push.

ws-rush committed:
rename to pocket-vue, switch from JSR to npm

ws-rush committed:
release: v1.2.1

ws-rush committed:
fix: correct @vitest/browser-playwright version specifier in package.json

ws-rush committed:
docs: update homepage url to pocket-vue.pages.dev

ws-rush committed:
feat(docs): add compact playground with shareable examples

- add portable playground page with live preview and code editor
- move docs examples into playground presets from examples and tests
- use CDN imports for playground snippets
- refine compact toolbar and pane styling
- remove old examples docs pages

ws-rush committed:
fix(docs): serve standalone playground without VitePress preload warnings

ws-rush committed:
docs: update logo file extension from png to jpg

ws-rush committed:
docs!: remove non-functional $data global and clean up legacy files

BREAKING: remove $data global variable that was documented but never
actually worked in expressions, causing ReferenceError at runtime.

- Delete docs/globals/data.md
- Remove $data nav link from docs sidebar config
- Remove $data from api-reference and v-effect docs
- Remove legacy examples/ and tests/ HTML files

ws-rush committed:
docs: improve README and add CONTRIBUTING guide

- Rewrite README with clearer project description, quick start,
  CDN/npm install instructions, and feature table
- Add CONTRIBUTING.md with dev setup, PR process, release workflow,
  and code style guidelines
- Fix npm publish workflow to include --access public flag

ws-rush committed:
docs: update documentation URL to pocket-vue.wusaby.com

ws-rush committed:
docs: fix docs accuracy and navigation

- add Vue dependency required for VitePress builds
- align API docs with runtime types and cleanup behavior
- clarify refs, v-bind, and v-effect behavior
- add Vue compatibility guide
- reorganize docs navigation and sidebar

ws-rush committed:
feat(docs): replace self-hosted playground with MyLab share links

Replace the custom in-page playground editor (iframe + code editor app)
with a directory page linking each of the 25 examples to a ready-to-edit
mylab.wsm.one/play URL.

- Remove docs/public/playground/ (index.html editor app, examples.json)
- Remove buildEnd hook and unused fs/path imports from VitePress config
- Fix TodoMVC: wrap localStorage in try/catch for MyLab sandbox
  (lacks allow-same-origin, so localStorage throws SecurityError)
- Fix Markdown: convert external <script src> tags (marked, lodash) to
  ESM imports so module execution waits for dependencies to load

ws-rush committed:
Merge pull request #17 from ws-rush/feat/playground-mylab-links

feat(docs): replace self-hosted playground with MyLab share links

ws-rush committed:
docs(playground): update MyLab playground domain to byrush.me

ws-rush committed:
Merge pull request #18 from ws-rush/feat/playground-mylab-links

docs(playground): update MyLab playground domain to byrush.me

ws-rush committed:
docs: update domain to pocket-vue.byrush.me

ws-rush committed:
docs: rename playground to examples

ws-rush committed:
Merge pull request #19 from ws-rush/feat/update-domain

docs: update domain to pocket-vue.byrush.me
